text/3057-boxed-macro.md

```diff
@@ -0,0 +1,114 @@
+- Feature Name: `boxed_macro` 
+- Start Date: 2020-12-29
+- RFC PR: [rust-lang/rfcs#3057](https://github.com/rust-lang/rfcs/pull/3057)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Introduces the `boxed` macro, which allows one to construct data directly on the heap. 
+This provides an alternative method to box syntax that does not rely on magic syntax. In addition, 
+if box syntax is significantly revised in the future, these changes could be applied to the `boxed` macro to 
+prevent ecosystem breakage.
+
+# Motivation
+[motivation]: #motivation
+
+At the moment, the only safe way of constructing data directly on the heap in Rust is via box syntax.
+The [`vec`](https://doc.rust-lang.org/src/alloc/macros.rs.html#37-47) macro is implemented in terms of it,
+and several other parts of the [standard library](https://doc.rust-lang.org/src/alloc/sync.rs.html#314-318)
+use it. 
+[`Box::new`](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.new) is the stable alternative to this;
```

Comment:
Can we think of any examples that aren't already served by the `vec!` macro that are likely to be big enough to overflow the stack?

This function for example:

```rust
fn big_boxed_slice(init: u8) -> Box<[u8]> {
    Box::new([init; 10485760])
}
```

allocates a 10MiB array, and (currently) overflows in both debug and release mode.

The alternative with `box expr` (or `boxed!`) would be:

```rust
fn big_boxed_syntax_slice(init: u8) -> Box<[u8]> {
    boxed!([init; 10485760])
}
```

which doesn't overflow, but is equivalent to the already stable:

```rust
fn big_vec(init: u8) -> Box<[u8]> {
    vec![init; 10485760].into_boxed_slice()
}
```

[Playground link for exploring some asm](https://play.integer32.com/?version=nightly&mode=release&edition=2018&gist=0821b7f9ed92d872a73369093e352f33)